### PR TITLE
Invoice reconcile

### PR DESCRIPTION
**Pull request by pepcus-adityatripathi**
Hi @deepak-nigam-pepcus  please review the code changes.

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -6,14 +6,31 @@
...
 """
 
-from fastapi import FastAPI, Header
+from fastapi import FastAPI, Header,File,UploadFile
 from pydantic import BaseModel
 from utils.pinecone_vectorstore import build_pinecone_index
 from utils.chat import get_chat_response
 from auth import is_authorized_request
+import shutil
+from pathlib import Path
```
Comment (deepak-nigam-pepcus) [inline app.py]: what is the use of this module?

File: app.py [hunk 2]
```diff
@@ -6,14 +6,31 @@
...
+from utils.invoice_processing import get_file_type,extract_text_based_on_file_type,fetch_invoice_details
+
 
 app = FastAPI()
 
+# Define the directory to store uploaded files
+UPLOAD_DIR = Path(__file__).resolve().parent / "uploads"
+UPLOAD_DIR.mkdir(exist_ok=True)
+# This example allows all origins, adjust as needed for production
+app.add_middleware(
```
Comment (deepak-nigam-pepcus) [inline app.py:28]: What is the use of these middlewares?

File: app.py [hunk 3]
```diff
@@ -6,14 +6,31 @@
...
+import shutil
+from pathlib import Path
+from fastapi.middleware.cors import CORSMiddleware
+from utils.invoice_processing import get_file_type,extract_text_based_on_file_type,fetch_invoice_details
+
 
 app = FastAPI()
 
+# Define the directory to store uploaded files
+UPLOAD_DIR = Path(__file__).resolve().parent / "uploads"
```
Comment (deepak-nigam-pepcus) [inline app.py]: can we move this to config file?

File: app.py [hunk 4]
```diff
@@ -33,6 +50,27 @@ def create_index(company: str, authorization: str = Header(None, convert_undersc
     is_authorized_request(auth=authorization)
     build_pinecone_index(company)
 
+@app.post("/api/invoice")
+def upload_invoice(file: UploadFile = File(...),authorization: str = Header(None, convert_underscores=True)):
+    is_authorized_request(auth=authorization)
+    file_path = UPLOAD_DIR / file.filename
+    with open(file_path, "wb") as buffer:
+        shutil.copyfileobj(file.file, buffer)
+    file_type=get_file_type(file.filename)
```
Comment (deepak-nigam-pepcus) [inline app.py:68]: please give proper spacing in code for better readability

File: utils/invoice_processing.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        print("======= {} =======".format(file_type))
+        text_extraction_function = extraction_functions[file_type]
+        text = text_extraction_function(file_path)
+        extracted_text=text
+        return text
+    else:
+        print("Unsupported file type:", file_type)
+        return None
+
+def fetch_invoice_details(text):
```
Comment (deepak-nigam-pepcus) [inline utils/invoice_processing.py:12]: We can extract out this prompt in a variable, as of now the prompt is tightly coupled with the method.

File: utils/sql.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+Date created: 2024-06-06
+License: MIT License
+Description: This file contains SQL related functions.
+"""
+
+from langchain_community.utilities import SQLDatabase
+from config.config import pg_db_uri
+
+def execute_query(query):
+    print("=======going to execute below query===========", query)
```
Comment (deepak-nigam-pepcus) [inline utils/sql.py]: Remove this print and use logger

File: app.py (hunk 1 above)
Comment (pepcus-adityatripathi) [reply app.py]: using shutil to create a file from file object that we are recieving from Front end .

File: app.py (hunk 2 above)
Comment (pepcus-adityatripathi) [reply app.py:28]: Was getting cross origin error(405) while sending file to back end that is why used these configuration.

File: app.py (hunk 3 above)
Comment (pepcus-adityatripathi) [reply app.py]: Sure!

File: app.py (hunk 4 above)
Comment (pepcus-adityatripathi) [reply app.py:68]: Ok!

Comment (pepcus-adityatripathi) [reply app.py:68]: i will add comments as well!

File: utils/sql.py (hunk 6 above)
Comment (pepcus-adityatripathi) [reply utils/sql.py]: Ok!

File: app.py (hunk 1 above)
Comment (deepak-nigam-pepcus) [reply app.py]: The question is for 'pathlib' module, not for 'shutil'

**Pull request closed without merging**